subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -4628,28 +4534,129 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
 	}
 }
 
-/* LwM2M main work loop */
+static void create_notifications(sys_slist_t *observer_list,
+				 const int64_t timestamp)
+{
+	struct observe_node *obs;
+	int rc;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(observer_list, obs, node) {
+		if (obs->event_timestamp > obs->last_timestamp &&
+		    timestamp > obs->last_timestamp +
+		    MSEC_PER_SEC * obs->min_period_sec) {
+			rc = generate_notify_message(obs, true);
+			if (rc == -ENOMEM) {
+				/* no memory/messages available, retry later */
+				return;
+			}
+			obs->last_timestamp = timestamp;
+		} else if (timestamp >
+			   obs->last_timestamp +
+			   MSEC_PER_SEC * obs->max_period_sec) {
+			rc = generate_notify_message(obs, false);
+			if (rc == -ENOMEM) {
+				/* no memory/messages available, retry later */
+				return;
+			}
+			obs->last_timestamp = timestamp;
+		}
+	}
+}
 
-static void socket_receive_loop(void)
+static int socket_recv_message(struct lwm2m_ctx *client_ctx)
 {
 	static uint8_t in_buf[NET_IPV6_MTU];
-	static struct sockaddr from_addr;
 	socklen_t from_addr_len;
 	ssize_t len;
-	int i;
+	static struct sockaddr from_addr;
+
+	from_addr_len = sizeof(from_addr);
+	len = recvfrom(client_ctx->sock_fd, in_buf, sizeof(in_buf) - 1,
+		       MSG_DONTWAIT, &from_addr, &from_addr_len);
+
+	if (len < 0) {
+		if (errno == EAGAIN || errno == EWOULDBLOCK) {
+			return -errno;
+		}
+
+		LOG_ERR("Error reading response: %d", errno);
+		if (client_ctx->fault_cb != NULL) {
+			client_ctx->fault_cb(errno);
+		}
+		return -errno;
+	}
+
+	if (len == 0) {
+		LOG_ERR("Zero length recv");
+		return 0;
+	}
+
+	in_buf[len] = 0U;
+	lwm2m_udp_receive(client_ctx, in_buf, len, &from_addr, handle_request);
+
+	return 0;
+}
+
+static int socket_send_message(int sock_fd)
+{
+	for (int i = 0; i < CONFIG_LWM2M_ENGINE_MAX_MESSAGES; i++) {
+		if (messages[i].ctx && messages[i].send &&
+		    messages[i].ctx->sock_fd == sock_fd) {
+			return lwm2m_send_message(&messages[i]);
+		}
+	}
+
+	return 0;
+}
+
+static void socket_reset_pollfd_events(void)
+{
+	for (int i = 0; i < sock_nfds; ++i) {
+		sock_fds[i].events = POLLIN;
+		sock_fds[i].revents = 0;
+	}
+	for (int i = 0; i < CONFIG_LWM2M_ENGINE_MAX_MESSAGES; ++i) {
```

Comment:
Up for discussion, but I'd rather see `k_fifo` being used here, instead of looping over all allocated LwM2M messages every time. We could keep `k_fifo` per lwm2m_context (and thus socket) and then:
1. Put a message to the FIFO in `lwm2m_send_message_async()`,
2. Verify if the FIFO IS empty here, to decide whether to add POLLOUT,
3. Retrieve the message from the FIFO in `socket_send_message()`

All that is needed to make `lwm2m_message` FIFO ready is to add `void *fifo_reserved` at the beginning of the structure.

Reply:
Yeah, that's something I was hesitant about as well. I decided to go with the simple stupid approach first, and it seemed "fast enough" :tm: to me, but I see your point.

Reply:
I think a fifo might even be a little heavy-handed, since that's primarly used to synchronize between threads, but all producers are running in the same thread, so a slist might be enough. I'm curious what your opinion on that would be.

Reply:
I chose to go with a slist for now.